### PR TITLE
NewProjectCI 

### PR DESCRIPTION
**Pull request by Filmaluco**
a improved kotlin version of the bash script aiming to solve issue #9

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,10 +33,9 @@
 
 # Getting started
 
-### Script 
+### NewProjectCI 
 1. Download this repository extract and open the folder
-2. Run `newproject.sh` script to create a copy with the name you want 
-    * Example: `./newproject.sh -p com.example.project -t Project` 
+2. Run `java -jar NewProjectCI` script to create a copy with the name you want
```
Comment (sdsantos) [inline README.md:38]: Do we really need a `jar`, can't we compile and run just in time? Just so users don't have to trust a random "executable".

File: NewProjectCI/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+version = "1.0-SNAPSHOT"
+
+application {
+    mainClass.set("Main.kt")
+}
+
+repositories {
+    mavenCentral()
+}
+// https://stackoverflow.com/questions/26469365/building-a-self-executable-jar-with-gradle-and-kotlin
```
Comment (sdsantos) [inline NewProjectCI/build.gradle.kts:18]: It's not clear what this comment refers to.

File: NewProjectCI/settings.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+
```
Comment (sdsantos) [inline NewProjectCI/settings.gradle.kts:1]: Lines, empty lines, you love them 😄

File: NewProjectCI/src/main/kotlin/Main.kt [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    private val verbose by option().flag()
+
+    override fun run() = runBlocking {
+        logger.enabled = verbose
+
+        ProgressBar("New Project", 100).use { pb ->
+            pb.extraMessage = "Coping template"
+            sleep(100)
+            val (numberOfFilesToCopy, stateOfCopiedFiles) = copyMainTemplate()
+            var stepBy: Long = (numberOfFilesToCopy/20).toLong()
```
Comment (sdsantos) [inline NewProjectCI/src/main/kotlin/Main.kt:33]: This `20` that keeps appearing looks random.

File: NewProjectCI/src/main/kotlin/Main.kt [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        .check("(i.e. MyApp)") { it.matches(titleRegex) }
+
+    private val verbose by option().flag()
+
+    override fun run() = runBlocking {
+        logger.enabled = verbose
+
+        ProgressBar("New Project", 100).use { pb ->
+            pb.extraMessage = "Coping template"
+            sleep(100)
```
Comment (sdsantos) [inline NewProjectCI/src/main/kotlin/Main.kt:31]: Another magic number that pops up regularly, could be a constant if it's meant to be the same. And are we worried the steps are too fast? 😅

File: NewProjectCI/src/main/kotlin/Main.kt [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            stepBy = (20/numberOfSourceSets).toLong()
+            flowRenaming.collect {
+                logger.d(it)
+                pb.stepBy(stepBy)
+            }
+
+            pb.extraMessage = "Replacing data in files"
+            sleep(100)
+            replaceInFiles().collect {
+                if(it.isNotEmpty()) {
```
Comment (sdsantos) [inline NewProjectCI/src/main/kotlin/Main.kt:66]: Kotlin lint could give out a hand here.

File: NewProjectCI/src/main/kotlin/utils/FileUtils.kt [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        } else {
+            if (destDir.exists()) {
+                if (!destDir.isDirectory) {
+                    throw IOException("Destination '$destDir' exists but is not a directory")
+                }
+            } else if (!destDir.mkdirs() && !destDir.isDirectory) {
+                throw IOException("Destination '$destDir' directory cannot be created")
+            }
+            if (!destDir.canWrite()) {
+                throw IOException("Destination '$destDir' cannot be written to")
```
Comment (sdsantos) [inline NewProjectCI/src/main/kotlin/utils/FileUtils.kt:75]: If you are throwing you don't need to nest all the ifs.

File: NewProjectCI/src/main/kotlin/utils/Logger.kt [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        print("\n[$type] $dateTime : $message")
+    }
+
+    enum class LogType {
+        Debug
+    }
+
+    companion object {
+        val logger = Logger()
+    }
```
Comment (sdsantos) [inline NewProjectCI/src/main/kotlin/utils/Logger.kt:29]: Use `object` instead for kotlin singletons.

File: NewProjectCI/src/main/kotlin/utils/FileUtils.kt [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+fun File.copyDirectory(destDir: File): Pair<Int, MutableSharedFlow<String>> {
+    if (!isDirectory) {
+        throw IOException("Source '$this' exists but is not a directory")
+    } else if (canonicalPath == destDir.canonicalPath) {
+        throw IOException("Source '$this' and destination '$destDir' are the same")
+    } else {
+        val fileCopyFlow = MutableSharedFlow<String>()
+        val numberOfFiles = FileUtils.listFiles(this, FileFilterUtils.trueFileFilter(), TrueFileFilter.INSTANCE).size
+
+        CoroutineScope(Dispatchers.IO).launch {
```
Comment (sdsantos) [inline NewProjectCI/src/main/kotlin/utils/FileUtils.kt:49]: Why do some methods launch coroutines, some switch coroutines, and others don't care? Shouldn't that be handled instead by the main method or at least coherent across these methods?

File: NewProjectCI/src/main/kotlin/utils/Logger.kt [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+        val c = Calendar.getInstance()
+        val simpleDateFormat = SimpleDateFormat("dd-MMM-yyyy hh-mm-ss")
+        val dateTime = simpleDateFormat.format(c.time)
+        print("\n[$type] $dateTime : $message")
+    }
+
+    enum class LogType {
+        Debug
+    }
```
Comment (sdsantos) [inline NewProjectCI/src/main/kotlin/utils/Logger.kt:25]: Not very meaningful for now.

**Review by sdsantos — CHANGES_REQUESTED** [review]
Tried running the jar:
> java -jar NewProjectCI
```
Error: Invalid or corrupt jarfile NewProjectCI
```

File: NewProjectCI/build.gradle.kts (hunk 2 above)
Comment (Filmaluco) [reply NewProjectCI/build.gradle.kts:18]: oh this was a debug issue that I solve

File: NewProjectCI/src/main/kotlin/Main.kt (hunk 4 above)
Comment (Filmaluco) [reply NewProjectCI/src/main/kotlin/Main.kt:33]: 100% I have 5 tasks. 20% for each I will make it more verbose and scalable

File: NewProjectCI/src/main/kotlin/Main.kt (hunk 5 above)
Comment (Filmaluco) [reply NewProjectCI/src/main/kotlin/Main.kt:31]: yeah, if the steps are too fast the progress bar can't keep up

File: NewProjectCI/src/main/kotlin/utils/FileUtils.kt (hunk 7 above)
Comment (Filmaluco) [reply NewProjectCI/src/main/kotlin/utils/FileUtils.kt:75]: This was from a library so I kept it as it was, I will clean it

**Pull request closed without merging**